### PR TITLE
Use Distributed's new Computation API

### LINKED ISSUE #7650: `dumps_task` in `SimpleShuffleLayer` and `BroadcastJoinLayer` unpack
Just a TODO that `SimpleShuffleLayer` and `BroadcastJoinLayer` both use `dumps_task` in their `__dask_distributed_unpack__` methods, which leads to pickling on the scheduler, which we want to avoid.

Additionally, could this manual use of `pickle.dumps` in `BroadcastJoinLayer.__dask_distributed_pack__` lead to a double-pickle scenario, where when the task actually gets executed, the kwarg values are bytestrings, not the actual unpickled values: https://github.com/dask/dask/blob/b3a86469cb339767617c03c05e60db101ceb177a/dask/layers.py#L663-L671

Related to https://github.com/dask/distributed/pull/4699.

cc @rjzamora @madsbk

### LINKED ISSUE #7777: `dask.dataframe.read_csv('./filepath/*.csv')` returning tuple
**What happened**:
Loading a dataframe seemingly returned a tuple, rather than a `dask.dataframe`, as an exception was thrown:
`AttributeError: 'tuple' object has no attribute 'sample' `

**What you expected to happen**:
I expected for the code below to return a `pandas.DataFrame` with the correlations that I'm looking for!

**Minimal Complete Verifiable Example**:

```python
import dask.dataframe as daskdf
from dask.distributed import Client

client = Client(memory_limit='4GB', processes=False)

raw_df = daskdf.read_csv(os.path.join(input_file_path, '*.csv'))
df = raw_df.sample(frac=0.01).drop(['gaugeid', 'time', 'input', 'labels'], 1)
correlations = df.corr().compute()
```

**Anything else we need to know?**:
The example runs fine on my local machine (Windows 10, Dask 2021.1.1, Python 3.8.5), it is just failing when run in containerised compute provided by Azure.

The full traceback is here:
```
Traceback (most recent call last):
  File "correlation_analysis.py", line 43, in <module>
    correlations = df.corr().compute()
  File "/azureml-envs/azureml_datastore/lib/python3.8/site-packages/dask/base.py", line 285, in compute
    (result,) = compute(self, traverse=False, **kwargs)
  File "/azureml-envs/azureml_datastore/lib/python3.8/site-packages/dask/base.py", line 567, in compute
    results = schedule(dsk, keys, **kwargs)
  File "/azureml-envs/azureml_datastore/lib/python3.8/site-packages/distributed/client.py", line 2673, in get
    results = self.gather(packed, asynchronous=asynchronous, direct=direct)
  File "/azureml-envs/azureml_datastore/lib/python3.8/site-packages/distributed/client.py", line 1982, in gather
    return self.sync(
  File "/azureml-envs/azureml_datastore/lib/python3.8/site-packages/distributed/client.py", line 853, in sync
    return sync(
  File "/azureml-envs/azureml_datastore/lib/python3.8/site-packages/distributed/utils.py", line 354, in sync
    raise exc.with_traceback(tb)
  File "/azureml-envs/azureml_datastore/lib/python3.8/site-packages/distributed/utils.py", line 337, in f
    result[0] = yield future
  File "/azureml-envs/azureml_datastore/lib/python3.8/site-packages/tornado/gen.py", line 762, in run
    value = future.result()
  File "/azureml-envs/azureml_datastore/lib/python3.8/site-packages/distributed/client.py", line 1847, in _gather
    raise exception.with_traceback(traceback)
  File "/azureml-envs/azureml_datastore/lib/python3.8/site-packages/dask/dataframe/methods.py", line 352, in sample
    return df.sample(random_state=rs, frac=frac, replace=replace) if len(df) > 0 else df
AttributeError: 'tuple' object has no attribute 'sample'
```

**Environment**:

- Dask version: 2021.6.0
- Python version: 3.8.1
- Operating System: Linux
- Install method (conda, pip, source): conda

### PR DESCRIPTION
**Pull request by madsbk**
Depend on https://github.com/dask/distributed/pull/4923
- [ ] Closes #7650
- [ ] Closes #7777
- [ ] Tests added / passed
- [ ] Passes `black dask` / `flake8 dask` / `isort dask`

### CODE REVIEW

**Comment by mrocklin** [comment]
@rjzamora I imagine that you're familiar with this, but it would be good to get your thoughts from the perspective of writing layers.

File: dask/blockwise.py [hunk 1]
```diff
@@ -1076,28 +1055,11 @@ def make_blockwise_graph(
...
         else:
-            if kwargs:
-                val = (apply, func, args, kwargs2)
-            else:
-                args.insert(0, func)
-                val = tuple(args)
-            # May still need to serialize (if concatenate=True)
-            dsk[out_key] = to_serialize(val) if deserializing else val
+            val = (func,) + tuple(args)
+        dsk[out_key] = val
```
Comment (rjzamora) [inline dask/blockwise.py:1062]: nice :)

File: dask/highlevelgraph.py [hunk 2]
```diff
@@ -346,7 +346,7 @@ def __dask_distributed_pack__(
           - Serializable by msgpack (notice, msgpack converts lists to tuples)
           - All remote data must be unpacked (see unpack_remotedata())
           - All keys must be converted to strings now or when unpacking
-          - All tasks must be serialized (see dumps_task())
+          - All tasks must be typesetted (see typeset_dask_graph())
```
Comment (rjzamora) [inline dask/highlevelgraph.py:349]:
```suggestion
          - All tasks must be typeset (see typeset_dask_graph())
```
I have no idea if "typeset" is correct here, but I have a feeling that "typesetted" is not a word.

File: dask/blockwise.py [hunk 3]
```diff
@@ -531,28 +532,7 @@ def __dask_distributed_pack__(
...
-        # not cache the function on the scheduler or worker (or warn if there are
-        # large objects being passed into the graph).  However, in the future,
-        # single-pass serialization improvements should allow us to remove this
-        # special logic altogether.
-        func = (
-            to_serialize(dsk[0])
-            if (self.concatenate or inline_tasks)
-            else dumps_function(dsk[0])
-        )
+        func = PickledCallable.serialize(dsk[0])
```
Comment (rjzamora) [inline dask/blockwise.py:535]: This is nice, and it reminds me that we should probably abandon the use of [CallableLazyImport](https://github.com/dask/dask/blob/965e35fe5a9423f550991bde08116a4984498e68/dask/layers.py#L23) in existing and future `Layer` definitions.

**Review by rjzamora — COMMENTED** [review]
This and [dask/distributed#4923](https://github.com/dask/distributed/pull/4923) are looking really nice to me - Thanks for the work here Mads!

**Comment by rjzamora** [comment]
Note that we probably need test coverage for #7777 before we consider it officially "closed"

File: dask/highlevelgraph.py (hunk 2 above)
Comment (madsbk) [reply dask/highlevelgraph.py:349]: > I have a feeling that "typesetted" is not a word

Heheh, I think you are right! :)

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Comment by quasiben** [comment]
add to allowlist

**Pull request closed without merging**